### ISSUE TITLE
FR: Support for VARLEN fields with different size

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gixsql#54]
The size of sub-items to store the length of variable-length-fields can be currently chosen at compile-time like this:

    #ifdef USE_VARLEN_32
    #define VARLEN_LENGTH_PIC		"9(05) BINARY"
    #define VARLEN_LENGTH_SZ		4
    #define VARLEN_LENGTH_T			uint32_t
    #define VARLEN_BSWAP			COB_BSWAP_32
    #else
    #define VARLEN_LENGTH_PIC		"9(4) BINARY"
    #define VARLEN_LENGTH_SZ		2
    #define VARLEN_LENGTH_T			uint16_t
    #define VARLEN_BSWAP			COB_BSWAP_16
    #endif

The problem is that short/long field length indicators cannot be currently mixed, so in practice there is no difference between `VARCHAR` and `LONG VARCHAR`, provided that you define `USE_VARLEN_32` (except at declaration time, where I think `LONG VARCHAR` is still not supported, so you have to use standard `VARCHAR`).

_Originally posted by @mridoni in https://github.com/mridoni/gixsql/issues/38#issuecomment-1117389355_

So things to do:

* support both 2 and 4 byte sizes in the runtime together, likely adding an internal type
* adjust the preprocessor to insert the matching type when explicit specified via `EXEX SQL VAR`  and to use the 4byte type for `VARYING` groups which extend the 2byte size

**Comment by GitMensch** [comment]
Rechecked: OCESQL uses the following:

```c
#define OCDB_VARCHAR_HEADER_BYTE 4 // size of PIC S9(8) COMP-5`
```

So I do wonder: Shouldn't USE_VARLEN_32 be the default as long as there's a fixed varlen only?

**Comment by mridoni** [comment]
> So I do wonder: Shouldn't `USE_VARLEN_32` be the default as long as there's a fixed varlen only?

Yes, that's probably for the better, I'll add this to the next batch of fixes.

**Comment by GitMensch** [comment]
Should the "variable" support instead of GixSQL compile time be in the backlog or planned for a version?
For now I have no need for it, just want to get some overview where GixSQL stands and what issues are "in progress".

**Comment by mridoni** [comment]
This will be partially solved in v1.0.21dev (to be released in about a week). It will be possible to specify the size of the length indicator both during preprocessing and at runtime (with an environment variable). It is a stopgap but at least it allows for different indicator field sizes without having to recompile (or use a separate version of) GixSQL. This will be definitely fixed by adding a flag to the ones currently passed to define SQL parameters (along with type, size, etc.) but I don't know if it will be possible for v1.0.21 "final".
